### DISCUSSION TITLE
[Question][Hw1] Necessity of `incoming_mem`

### DISCUSSION THREAD

**Discussion by ironore15** [prosyslab-classroom/cs424-program-reasoning#411] [Q&A] [answered]
### Your Name

민승기

### Question

I'm trying to get a deeper understanding of `transfer`, and I don't get it why `prev` should be `(Llvm.llbasicblock * Memory.t option)`.

As far as I understand, `transfer` function is always called by previous instruction, and the argument `memory` should be automatically set with `incoming_mem`. Thus, they are same and `prev` is okay with just `(Llvm.llbasickblock) option`.

Am I wrong? or is this design just for future expansion?

**Comment by bongjunj** [comment, marked as answer]
According to the semantics of phi nodes, there are cases where `incoming_mem` is not identical to `memory`.

**Reply by ironore15** [reply]
After finding some semantic definitions of phi node, I think I still don't get it how it happens.

Is there any other instruction except `br` that can enter to the phi node? Or the situation above also can happen even with `br` instruction?

**Reply by bongjunj** [reply]
Yes, a block can contain more than one phi nodes.

**Reply by ironore15** [reply]
Now I fully understood. Thanks

**Comment by spearo2** [comment]
I am closing this discussion. Thanks @bonjune for the contribution.
